### ISSUE TITLE
Extend the Kabanero Collection CRD to allow multiple stack versions to be active simultaneously.

### ISSUE THREAD

**Issue by stephenkinder** [kabanero-io/kabanero-operator#288]
Document the design for the Kabanero Collection CRD schema to support multiple versions of the same Kabanero Collection to be active at the same time.

**Comment by kaczyns** [comment]
I'm proposing the following CR instance structure, which maintains all
of the fields that exist prior to now, as well as the new array that
will allow us to enable multiple versions of the same collection at
the same time:

```yaml
apiVersion: kabanero.io/v1alpha1
kind: Collection
metadata:
  name: java-microprofile
  namespace: kabanero
spec:
  name: java-microprofile
  desiredState: active
  repositoryUrl: https://whatever.com/blah/blah
  version: 0.2.15
  versions:
  - version: 0.2.15
    desiredState: active
    repositoryUrl: https://whatever.com/blah/blah
  - version: 0.2.16
    desiredState: active
    repositoryUrl: https://whatever.com/blah/blah
  - version: 0.2.17
    desiredState: active
    repositoryUrl: https://whatever.com/blah/blah
status:
  activeLocation: https://whatever.com/blah/blah
  activePipelines:
  - name: default
    digest: abc123
    url: https://whatever.com/common.pipeline.default.tar.gz
    activeAssets:
    - assetDigest: efg456
      assetName: ./build-pipeline.yaml
      status: active
    - assetDigest: try934
      aseetName: ./some-other.yaml
      status: active
    activeVersion: 0.2.15
    images:
    - id: java-microprofile
      image: kabanero/java-microprofile:0.2
    status: active
  activeVersions:
  - location: https://whatever.com/blah/blah
    pipelines:
    - name: default
      digest: abc123
      url: https://whatever.com/common.pipeline.default.tar.gz
      assets:
      - assetDigest: efg456
        assetName: ./build-pipeline.yaml
        status: active
      - assetDigest: try934
        aseetName: ./some-other.yaml
        status: active
      version: 0.2.15
      images:
      - id: java-microprofile
        image: kabanero/java-microprofile:0.2
      status: active
  - location: https://whatever.com/blah/blah
    pipelines:
    - name: default
      digest: abc123
      url: https://whatever.com/common.pipeline.default.tar.gz
      assets:
      - assetDigest: efg456
        assetName: ./build-pipeline.yaml
        status: active
      - assetDigest: try934
        aseetName: ./some-other.yaml
        status: active
      version: 0.2.16
      images:
      - id: java-microprofile
        image: kabanero/java-microprofile:0.2
      status: active
  - location: https://whatever.com/blah/blah
    pipelines:
    - name: default
      digest: abc123
      url: https://whatever.com/common.pipeline.default.tar.gz
      assets:
      - assetDigest: efg456
        assetName: ./build-pipeline.yaml
        status: active
      - assetDigest: try934
        aseetName: ./some-other.yaml
        status: active
      version: 0.2.17
      images:
      - id: java-microprofile
        image: kabanero/java-microprofile:0.2
      status: active
```

Note that only `spec.versions` and `status.activeVersions` are new.
The versions[0] duplicates the former singleton values, and
activeVersions[0] duplicates the former singleton status values.  This
mapping is required so that older clients can continue to function.
Our intention will be to introduce a new version of the CRD which
eliminates the singleton versions of these fields, and deprecate this
one.  I am proposing `v1alpha2` as the new version.

A CRD has a conversion strategy.  Currently in OCP 4.2 `none` is the
only supported strategy.  This means that the fields between CRD
versions need to line up exactly.  This is why the `v1alpha1` CRD has
to have exactly the same field names as the `v1alpha2` CRD will have.
In OCP 4.3, we will be able to use a conversion webhook to avoid this,
however it will be some time before OCP 4.3 is a prerequisite.  See:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md
and
https://v1-14.docs.kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#webhook-conversion

The resulting CRD is large and unwieldy but will allow products built
on Kabanero to continue to function while we make the conversion.  The
next version of Kabanero can introduce a v1alpha2 CRD which is served
but not stored, and only contains the new fields.  Then in a future
release of Kabanero, we can store the v1alpha2 version, and finally in
another future release of Kabanero we can remove the v1alpha1 CRD
altogether.

We will require an admission webhook (see #57) to enforce that the
singleton `spec.version` field does not contradict the array form
`spec.versions`.  There are two admission webhook types - validating
and mutating.  A mutating webhook can change values, while a
validating webhook can not.  I would propose that we have a mutating
web hook which makes sure that the singleton version represents the
first item in the versions array.  Then the validating webhook makes
sure that the singleton and the first item in the version array
match.  Things get tricky when the user updates the CR instance - see
the "Params" and "HeightInInches" examples in the `api_changes.md`
link above.  We will need to compare the "old" and "new" version of
the object to see what changed, and make sure the mapping between
`version` and `versions[0]` is consistent.

This comment does not address the desire (or requirement) to re-name
the collection resources (pipelines, tasks, etc) so that they can be
re-used across collection versions.  I'll make an additional comment
with a proposal on that.

**Comment by kaczyns** [comment]
This comment is going to try to address the collection / pipeline
artifact naming problem.  When we have multiple versions of the same
collection active, at the same time, they may reference different
versions of the collection / pipeline artifacts.  For example,
java-microprofile v0.2.15 might refer to pipeline zip v1, but
java-microprofile v0.2.16 might refer to pipeline zip v2.  Since the
artifacts in the pipeline zips will likely have the same name, and may
not be compatible with each other, we need to dis-ambiguate somehow.
Conversely, the two versions of the collection might use exactly the
same versions of the collection / pipeline artifacts, and in that
case, we do not want to deply the same artifact again with a different
name.  Taking it a bit further, there may be unrelated collections
that happen to use the same collection / pipeline artifacts, and it
would be nice to not deploy the same artifacts again with different
names in that case as well.

So lets tack the first 8 characters of the SHA256 on the end of the
artifact name.  The artifact will need to have a SHA for itself in the
`manifest.yaml` inside the pipeline zip.  This way, we can have
multiple versions of the same artifact.  Sure, it's going to get
tricky to figure out which pipeline should be used.  The best we can
do is print the name of the resource in the status.  That way tooling
etc can look for it there and pick the right resource.  If someone
wants to do a manual pipeline run, using the Tekton dashboard, they
will have to pound sand.  or something.

I think the collection reconciliation process is going to work
something like this:
1) Go thru the collection status, and make a data structure that
   contains every in-use artifact and how many times it is used (if
   it's used by two versions of the same collection, its use count
   gets bumped).
2) Process any collections that are transitioning to inactive, or are
   deleted from the version list.  Decrement the count the appropriate
   amount.  If the use count goes to zero, leave it in the list with a
   use count of zero, so that we know later we'll need to delete it.
3) Process any collections that are new or transitioning to active.
   Increment the count the appropriate amount.  For new resouces, add
   to the list and bump the count.
4) Go thru the list.  For any resources with a use count of 0, remove
   the collection object from the object ownership.  if the ownership
   is now empty, delete the resource.  For the rest of the resources,
   if the count is positive, try to GET the resource.  if it's not
   there, create it, setting the collection object as the owner.  We
   can discuss whether it's appropriate to apply the YAML again even
   if the resource is still there, to undo any changes the user may
   have made to the object.  I'm not sure if this is appropriate or
   necessary.

If the owner of the collection hub removes the support that we added
previously to prefix the collection / pipeline artifact names with the
name of the stack (ie java-microprofile-), then the number of
artifacts should be reduced greatly if the collections all share the
same pipelines.

This process is going to depend on having the files in the collection
hub cached, so that we're not constantly going to Github to fetch the same
collection pipeline zip over and over again.  See #147.  Ideally both
the index yaml and the pipeline zips are cached, along with their etag
values, and we'll have a set of reasonable time constraints like:
1) Don't check Github again if we've checked in the past 60 seconds.
2) Throw away cached data every X hours (not sure what X is).

This reconciliation process is going to be significantly different
that Kabanero 0.3.0, so I think #58 should also be a prerequisite
here.  We could also just dual-path the behavior using the associated
`kabanero` version to determine what to do.  I think in the end,
deploying the controller separately will be less work.

**Comment by kaczyns** [comment]
This issue should be used to tie together the work that was done in the referenced issues above.  Namely, modify the Collection CRD to include the version array, and update the collection controller to activate and/or deactivate the individual versions of the collection.  The mutating webhook should be modified to align the singleton version and the 0th element of the versions array, using the old and new objects to determine which field needs to change, and failing if it can't be determined what to do.  The validating webhook should make sure that the version and 0th element of the versions array are the same, again failing if they are not.

**Comment by kaczyns** [comment]
This has merged and will ship in Kabanero 0.5.0.
